### ISSUE TITLE
Error: RStudio not running Execution halted

### ISSUE THREAD

**Issue by tungmilan** [rstudio/rstudio#1738]
I got the following error while knitting a Rmd file:   
`Line 17 Error: RStudio not running Execution halted`

	---
	title: "Results"
	author: "ABC"
	output:
	  pdf_document:
		number_sections: yes
	  html_document:
		number_sections: yes
		df_print: paged
	geometry: margin=0.75in
	urlcolor: blue
	---

	- Initialization 

	```{r}
	# set working directory
	library(rstudioapi)
	rootDir <- dirname(rstudioapi::getActiveDocumentContext()$path)
	rootDir
	setwd(rootDir)

	# load library
	library(lubridate)  
	library(magrittr)   
	library(tidyverse)  
	```

## System info:
	Session info --------------------------------------
	 setting  value                       
	 version  R version 3.4.2 (2017-09-28)
	 system   x86_64, mingw32             
	 ui       RStudio (1.2.163)           
	 language (EN)                        
	 collate  English_United States.1252  
	 tz       America/Los_Angeles         
	 date     2017-11-10                  

	Packages ------------------------------------------
	 package    * version    date      
	 backports    1.1.1      2017-09-25
	 base       * 3.4.2      2017-09-28
	 compiler     3.4.2      2017-09-28
	 datasets   * 3.4.2      2017-09-28
	 devtools     1.13.3     2017-08-02
	 digest       0.6.12     2017-01-27
	 evaluate     0.10.1     2017-06-24
	 graphics   * 3.4.2      2017-09-28
	 grDevices  * 3.4.2      2017-09-28
	 htmltools    0.3.6      2017-04-28
	 knitr        1.17       2017-08-10
	 magrittr     1.5.0      2017-10-17
	 memoise      1.1.0      2017-04-21
	 methods    * 3.4.2      2017-09-28
	 Rcpp         0.12.13.2  2017-11-08
	 rmarkdown    1.6.0.9009 2017-11-08
	 rprojroot    1.2        2017-01-16
	 rstudioapi * 0.7        2017-09-07
	 stats      * 3.4.2      2017-09-28
	 stringi      1.1.5      2017-04-07
	 stringr      1.2.0      2017-02-18
	 tools        3.4.2      2017-09-28
	 utils      * 3.4.2      2017-09-28
	 withr        2.1.0.9000 2017-11-08
	 yaml         2.1.14     2016-11-12
	 source                             
	 CRAN (R 3.4.1)                     
	 local                              
	 local                              
	 local                              
	 CRAN (R 3.4.1)                     
	 CRAN (R 3.4.0)                     
	 CRAN (R 3.4.1)                     
	 local                              
	 local                              
	 CRAN (R 3.4.0)                     
	 CRAN (R 3.4.1)                     
	 Github (tidyverse/magrittr@0a76de2)
	 CRAN (R 3.4.0)                     
	 local                              
	 Github (RcppCore/Rcpp@f91cd2e)     
	 Github (rstudio/rmarkdown@8fce71b) 
	 CRAN (R 3.4.2)                     
	 CRAN (R 3.4.2)                     
	 local                              
	 CRAN (R 3.4.0)                     
	 CRAN (R 3.4.0)                     
	 local                              
	 local                              
	 Github (jimhester/withr@a9ebeb3)   
	 CRAN (R 3.3.2)

**Comment by dfalty** [comment]
I'm able to reproduce this on 1.1.383.

**Comment by jmcphers** [comment]
R Markdown documents are rendered in a separate R session that doesn't have access to the rstudioapi. The error you're seeing:

    RStudio not running

is from the rstudioapi, letting you know it can't talk to RStudio (which is expected, since it's not in the same R session!).

It looks like what you're trying to do is set the knit directory to the directory of the document. This is the default knit behavior, so it shouldn't be necessary to write any code to accomplish it. Can you tell us a little more about the intent behind the rstudioapi call?

**Comment by tungmilan** [comment]
> This is the default knit behavior, so it shouldn't be necessary to write any code to accomplish it.
I didn't know that. I only want to set working directory to the same as where the script is. Thanks @jmcphers !

**Comment by samaperrin** [comment]
Hi, I keep getting a very similar error while trying to knit an R-markdown document.

This chunk works fine...
````{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

database <- 'db_name'
server <- "server_name"
````

However when I get down to my next chunk, 

``` {r slope_connect}
source("../R/slope_dbconnect.R")
con <- slope_dbconnect(database, server)
```

I get the 'Rstudio not running Execution halted' message.

Any help here would be greatly appreciated.

**Comment by mstuart1** [comment]
I also get this error, but I'm trying to access an SQL database in a package and this error is when I am using devtools::check() on my functions.  
Example errors:
4: .local(drv, ...)
5: rstudioapi::askForPassword("Username")
6: callFun("askForPassword", prompt)
7: verifyAvailable()
8: stop("RStudio not running", call. = FALSE)

How do I create a package that accesses an SQL database with a popup window for password entry?

**Comment by kevinushey** [comment]
You need to wrap your usages of `rstudioapi` in `rstudioapi::isAvailable()`. (RStudio APIs are not available for the R processes launched during testing)

**Comment by jmcphers** [comment]
It'd be cool if they were, though. See https://github.com/rstudio/rstudio/issues/4490.

**Comment by mstuart1** [comment]
Hi Kevin,

Do you have any advice for how to get my function to work during package testing or even knitting an Rmarkdown document?

We are trying to create a package for our working group that accesses our SQL database using dplyr and dbplyr.

Currently, our function looks like this:

read_db <- function(db_name){
  con <- DBI::dbConnect(RMySQL::MySQL(),
                        dbname = db_name,
                        host = "amphiprion.deenr.rutgers.edu<http://amphiprion.deenr.rutgers.edu>",
                        username = rstudioapi::isAvailable(rstudioapi::askForPassword("Username")),
                        password = rstudioapi::isAvailable(rstudioapi::askForPassword("Password")))
  return(con)
}

And adding rstudioapi::isAvailable causes it to fail all the time.  It used to only fail during pkg testing and markdown knitting.

Any advice you can offer is hugely appreciated!
Cheers,
Michelle

On Jul 9, 2019, at 5:11 PM, Kevin Ushey <notifications@github.com<mailto:notifications@github.com>> wrote:


You need to wrap your usages of rstudioapi in rstudioapi::isAvailable(). (RStudio APIs are not available for the R processes launched during testing)

**Comment by kevinushey** [comment]
@mstuart1 You might want to check out the [askpass](https://cran.r-project.org/web/packages/askpass/index.html) package on CRAN, which might work in more contexts than just RStudio.

If that doesn't help, you might just need an alternate way to provide credentials -- e.g. by setting environment variables that can be read and used from your R session, or by reading from a file (sufficiently locked down permission-wise) that provides access.

**Comment by mstuart1** [comment]
Thanks Kevin!!  We had hoped rstudioapi was the solution to our issues.  I will check out askpass.  We used to reference a config file for the username and password but after one of those package installs that updates a bunch of dependent packages, we started receiving this error message:
Error in .local(drv, ...) :
Failed to connect to database: Error: Can't connect to local MySQL server through socket '/tmp/mysql.sock' (2)
In addition: Warning message:
In .local(drv, ...) : partial argument match of 'group' to ‘groups'


On Jul 17, 2019, at 5:36 PM, Kevin Ushey <notifications@github.com<mailto:notifications@github.com>> wrote:


@mstuart1<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmstuart1&data=02%7C01%7Cmichelle.stuart%40rutgers.edu%7C57601b57c018448fb75a08d70afed8b1%7Cb92d2b234d35447093ff69aca6632ffe%7C1%7C0%7C636989961977636626&sdata=NSMrB1Dsne8U8%2FVYUYWWR%2BoBF1rV5ll96uEOGC5ynE4%3D&reserved=0> You might want to check out the askpass<https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fcran.r-project.org%2Fweb%2Fpackages%2Faskpass%2Findex.html&data=02%7C01%7Cmichelle.stuart%40rutgers.edu%7C57601b57c018448fb75a08d70afed8b1%7Cb92d2b234d35447093ff69aca6632ffe%7C1%7C0%7C636989961977646619&sdata=WWnk3B3WGdu1fB9z4mrrBLwBYUmHP6E4S3G0slfSKN0%3D&reserved=0> package on CRAN, which might work in more contexts than just RStudio.

If that doesn't help, you might just need an alternate way to provide credentials -- e.g. by setting environment variables that can be read and used from your R session, or by reading from a file (sufficiently locked down permission-wise) that provides access.
